README.md
# Codiad Managed VMs Docker Image

![Screenshot of Codiad running on MVMs](/codiad-screenshot.png?raw=true "Screenshot of Codiad running on MVMs")


This repo defines a Docker Image that can be used to run Codiad on App Engine Managed VMs.

## Codiad

The image is using a heavily customized version of Codiad (https://github.com/Codiad/Codiad) which
resides in `codiad/third_party/codiad` directory.

## IDE Proxy

In order to run the IDE instances for developers in a cloud project, there is an IDE proxy running
on the managed VM which routes the requests to the individual Codiad containers for each developer
based on their email. The IDE proxy is written in nodejs. It is responsible for creating a Codiad
container for a user as soon as they connect to the IDE. Thereafter all of the request coming from
that user will be forwarded to the respective container. The authentication is done through
appengine. Proxy code can be found under `ide-proxy` directory.


## Included extras

- Kythe: used for source code indexing.
    Kythe is installed by downloading its .tar.gz image from the github repository.
    We try to keep up with latest release of Kythe from https://github.com/google/kythe/releases/.
- Shipshape: for providing lint messages in IDE.
    Shipshape is installed by downloading its command line program from their public GCS bucket.
- Web terminal: A customized version of term.js(https://github.com/chjj/term.js) is what we use in
    IDE to expose the underlying container. IDE is installed through a git submodule which resides
    in a Google cloud repository.
- Git appraise: Integrated with git-appraise for sending code reviews.
    See https://github.com/google/git-appraise for more information. Here is a screenshot of
    integration of this IDE with git appraise: ![Screenshot of a review in Codiad](/codiad-review-screenshot.png?raw=true "Screenshot of a review in Codiad")

## Required Google Cloud API

- Google Compute Engine
- Google Cloud Storage

## Optional Google Cloud API

- Cloud Debugger API

## Building

### Building a dev image to test your changes to IDE

In order to do so, please see the `README.md` file in `codiad` directory.

### Deploying to a Google Cloud project

Please see the instructions in `ide-proxy` directory.

## How does it work?

The IDE is deployed as a Managed VM (MVM) module into a cloud project. This module is responsible
for handling requests coming from different users of the IDE and dispatching them to the right
Codiad container. The general architecture of the system is as follows:

```
                       +      +----------------------------------+
                       |      |                                  |
User A +----------->   +----------------+      IDE Proxy         |
                       |      |         |                        |
                       |      |         |                        |
User B +----------->   +--------------------------------+        |
                       |      |         |               |        |
                       |      |         |               |        |
                       +      |         |               |        |
                      MVM     |         |               |        |
                      Auth    |         |               |        |
                              |         |               |        |
                              |    +----v------+  +-----v-----+  |
                              |    | User A    |  | User B    |  |
                              |    | Codiad    |  | Codiad    |  |
                              |    | Container |  | Container |  |
                              |    +-----------+  +-----------+  |
                              +----------------------------------+
```

IDE proxy manages the codiad containers. It is respobsible for creating(and recreating) them and
routing requests to them based on user's email address which is registered with the cloud project.

### Security
As can be seen in the above diagram, the authentication for each user is done by MVM. Users are
authenticated with their cloud credential for the cloud project. The IDE is accessed by all users
using the same secure URL, e.g. https://codiad-YOUR_CLOUD_PROJECT.appspot.com.

The Codiad containers are run by project's service account and not by user's credential. No user's
credential is stored anywhere in the container by the system.

Codiad containers for all users are run in the same machine and there is no security boundary
established for them. Hence a user `A`'s container could potentially access user `B`'s container.

## Running Codiad Locally
It is possible to run Codiad locally on a machine without using the IDE proxy and MVM hosting. In
order to do so, see the `README.md` file in `codiad` directory.


RELEASES.md
# Release Notes

## Upcoming release

Notable changes:
 - Bumping up the version of codemirror to 5.8.0.

Notable changes:
 - Pinning down the version of gcloud to 0.9.79.

## v0.2-alpha

Notable features:
 - Split editing: user will be able to have two editors side by side for split editing.
 - Scratch buffer: there will be a scratch buffer where user can use as a temporary space for
   editing.
 - File path panels at the top of editors.

Notable changes:
 - Bumping up the version of nodejs to 0.12.7.
 - Bumping up the version of socket.io to 1.3.7.
 - Bumping up the version of node-http-proxy to 1.11.2.

Notable Fixes:
 - Fixed the codiad network creation issue on GCP.
 - Fixed the scratch buffer change issue.
 - Fixed the shipshape run issue.

## v0.1-alpha

First alpha release of Appengine Codiad IDE




test/test.js
var webdriverio = require('webdriverio');
var expect = require('expect.js');
var assert = require('assert');
var should = require('should');

var options = {
  host: 'localhost',
  port: 4444,
  desiredCapabilities: {
    browserName: 'firefox'
  }
};

describe('Codiad Tests', function(){

  this.timeout(120000); // 2 minutes
  var client = {};

  before(function(done){
    client = webdriverio.remote(options);
    client.init(done);
  });

  it('Codiad Terminal',function(done) {
    client.url('http://localhost:10000/').
    getTitle(function(err, title) {
      assert(err === undefined);
      assert(title === 'Codiad');
    }).
    click('span.content-closed div.terminal-toggler').
    waitForVisible('#terminal-pane', 2000).waitForVisible('div.terminal', 2000).
    getText('div.terminal').
    then(function(text) {
      expect(text).to.contain("root's home directory is '/workspace/'.");
    }).
    call(done);
  });

  it('Codiad Settings',function(done) {
    client.url('http://localhost:10000/').
    click('div#settings-icon').
    waitForVisible('.settings-view', 2000).
    getText('table.settings').
    then(function(text) {
      expect(text).to.contain("Line Numbers");
    }).
    call(done);
  });

  it('Create File',function(done) {
    client.url('http://localhost:10000/').
    rightClick('a#project-root').
    waitForVisible('div#context-menu', 2000).
    click('=New File').
    waitForVisible('div#modal-content', 2000).
    setValue('form.codiad-form input[name="object_name"]', 'file.md').
    click('button*=Create').
    getText('a.file').
    then(function(value) {
      expect(value).to.contain("file.md");
    }).
    call(done);
  });

  after(function(done) {
    client.end(done);
  });
});


test/wdio.conf.js
// Copyright 2015 Google Inc. All Rights Reserved.

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//     http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

exports.config = {

    //
    // ==================
    // Specify Test Files
    // ==================
    // Define which test specs should run. The pattern is relative to the directory
    // from which `wdio` was called. Notice that, if you are calling `wdio` from an
    // NPM script (see https://docs.npmjs.com/cli/run-script) then the current working
    // directory is where your package.json resides, so `wdio` will be called from there.
    //
    specs: [
        './specs/**/*.js'
    ],
    // Patterns to exclude.
    exclude: [
        // 'path/to/excluded/files'
    ],
    //
    // ============
    // Capabilities
    // ============
    // Define your capabilities here. WebdriverIO can run multiple capabilties at the same
    // time. Depending on the number of capabilities, WebdriverIO launches several test
    // sessions. Within your capabilities you can overwrite the spec and exclude option in
    // order to group specific specs to a specific capability.
    //
    // If you have trouble getting all important capabilities together, check out the
    // Sauce Labs platform configurator - a great tool to configure your capabilities:
    // https://docs.saucelabs.com/reference/platforms-configurator
    //
    capabilities: [{
        browserName: 'chrome'
    }],
    //
    // ===================
    // Test Configurations
    // ===================
    // Define all options that are relevant for the WebdriverIO instance here
    //
    // Level of logging verbosity.
    logLevel: 'verbose',
    //
    // Enables colors for log output.
    coloredLogs: true,
    //
    // Saves a screenshot to a given path if a command fails.
    screenshotPath: './errorShots/',
    //
    // Set a base URL in order to shorten url command calls. If your url parameter starts
    // with "/", the base url gets prepended.
    baseUrl: '',
    //
    // Default timeout for all waitForXXX commands.
    waitforTimeout: 10000,
    //
    // Initialize the browser instance with a WebdriverIO plugin. The object should have the
    // plugin name as key and the desired plugin options as property. Make sure you have
    // the plugin installed before running any tests. The following plugins are currently
    // available:
    // WebdriverCSS: https://github.com/webdriverio/webdrivercss
    // WebdriverRTC: https://github.com/webdriverio/webdriverrtc
    // Browserevent: https://github.com/webdriverio/browserevent
    // plugins: {
    //     webdrivercss: {
    //         screenshotRoot: 'my-shots',
    //         failedComparisonsRoot: 'diffs',
    //         misMatchTolerance: 0.05,
    //         screenWidth: [320,480,640,1024]
    //     },
    //     webdriverrtc: {},
    //     browserevent: {}
    // },
    //
    // Framework you want to run your specs with.
    // The following are supported: mocha, jasmine and cucumber
    // see also: http://webdriver.io/guide/testrunner/frameworks.html
    //
    // Make sure you have the node package for the specific framework installed before running
    // any tests. If not please install the following package:
    // Mocha: `$ npm install mocha`
    // Jasmine: `$ npm install jasmine`
    // Cucumber: `$ npm install cucumber`
    framework: 'mocha',
    //
    // Test reporter for stdout.
    // The following are supported: dot (default), spec and xunit
    // see also: http://webdriver.io/guide/testrunner/reporters.html
    reporter: 'xunit',
    
    //
    // Some reporter require additional information which should get defined here
    reporterOptions: {
        //
        // If you are using the "xunit" reporter you should define the directory where
        // WebdriverIO should save all unit reports.
        outputDir: './reports/'
    },
    
    //
    // Options to be passed to Mocha.
    // See the full list at http://mochajs.org/
    mochaOpts: {
        ui: 'bdd'
    },
    
    //
    // =====
    // Hooks
    // =====
    // Run functions before or after the test. If one of them returns with a promise, WebdriverIO
    // will wait until that promise got resolved to continue.
    // see also: http://webdriver.io/guide/testrunner/hooks.html
    //
    // Gets executed before all workers get launched.
    onPrepare: function() {
        // do something
    },
    //
    // Gets executed before test execution begins. At this point you will have access to all global
    // variables like `browser`. It is the perfect place to define custom commands.
    before: function() {
        // do something
    },
    //
    // Gets executed after all tests are done. You still have access to all global variables from
    // the test.
    after: function(failures, pid) {
        // do something
    },
    //
    // Gets executed after all workers got shut down and the process is about to exit. It is not
    // possible to defer the end of the process using a promise.
    onComplete: function() {
        // do something
    }
};


test/codiad.spec.js
// Copyright 2015 Google Inc. All Rights Reserved.

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//     http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

var webdriverio = require('webdriverio');
var expect = require('expect.js');
var assert = require('assert');
var fs = require('fs');
var spawnSync = require('child_process').spawnSync;

var options = {
  host: 'localhost',
  port: 4444,
  desiredCapabilities: {
    browserName: 'firefox'
  }
};

var CODIAD_PORT = 10000;
var CODIAD = 'http://localhost:' + CODIAD_PORT;
var CODIAD_WS = '/tmp/ide-ws';
var WAIT_FOR_VISIBLE_MS = 10000;

describe('Codiad Tests', function(){

  this.timeout(120000); // 2 minutes
  var client = {};
  var seleniumId;
  var codiadId;

  before(function(done) {
    // Run the selenium image which is bundled with firefox:
    var output = spawnSync('docker',
                           ['run', '--net', 'host', '-d', '-P', 'selenium/standalone-firefox']);
    seleniumId = output.stdout.toString('utf8').trim();

    // Make sure the google/codiad image is present locally:
    output = spawnSync('docker', ['images', '-q', 'google/codiad']);
    assert(output.stdout.toString('utf8').trim() !== '', 'google/codiad image is not available!');

    // Run codiad
    output = spawnSync('docker', ['run', '-e', 'USER_EMAIL=youremail@company.com', '--privileged',
                                 '-d', '-v', CODIAD_WS + ':/usr/share/nginx/www/_', '-p',
                                 CODIAD_PORT + ':8080', 'google/codiad']);
    codiadId = output.stdout.toString('utf8').trim();

    // Initialize a selenium client
    client = webdriverio.remote(options);
    client.init(done);
  });

  // Make sure codiad loads:
  it('Codiad Loaded',function(done) {
    client.pause(10000).url(CODIAD).
    waitForVisible('div#file-manager', WAIT_FOR_VISIBLE_MS).
    getTitle(function(err, title) {
      assert(err === undefined);
      assert(title === 'Codiad');
    }).
    call(done);
  });

  // Make sure terminal appears:
  it('Codiad Terminal',function(done) {
    client.url(CODIAD).
    click('span.content-closed div.terminal-toggler').
    waitForVisible('#terminal-pane', WAIT_FOR_VISIBLE_MS).
    waitForVisible('div.terminal', WAIT_FOR_VISIBLE_MS).
    getText('div.terminal').
    then(function(text) {
      expect(text).to.contain("root's home directory is '/workspace/'.");
    }).
    call(done);
  });

  // Make sure settings dialog shows up:
  it('Codiad Settings',function(done) {
    client.url(CODIAD).
    click('div#settings-icon').
    waitForVisible('.settings-view', WAIT_FOR_VISIBLE_MS).
    getText('table.settings').
    then(function(text) {
      expect(text).to.contain("Line Numbers");
    }).
    call(done);
  });

  // Create a file:
  it('Create File',function(done) {
    client.url(CODIAD).
    rightClick('a#project-root').
    waitForVisible('div#context-menu', WAIT_FOR_VISIBLE_MS).
    click('=New File').
    waitForVisible('div#modal-content', WAIT_FOR_VISIBLE_MS).
    setValue('form.codiad-form input[name="object_name"]', 'file.md').
    click('button*=Create').
    getText('a.file').
    then(function(value) {
      expect(value).to.contain("file.md");
    }).
    call(done);
  });

  // Delete the created file:
  it('Delete File',function(done) {
    client.url(CODIAD).
    rightClick('a*=file.md').
    waitForVisible('div#context-menu', WAIT_FOR_VISIBLE_MS).
    click('=Delete').
    waitForVisible('div#modal-content', WAIT_FOR_VISIBLE_MS).
    click('button*=Delete').
    getText('div#file-manager').
    then(function(value) {
      expect(value).to.not.contain("file.md");
    }).
    call(done);
  });

  after(function(done) {
    client.end(done).then(function() {
      fs.unlink(CODIAD_WS + '/workspace/cloud-project/file.md', function(){});
      spawnSync('docker', ['stop', codiadId]);
      spawnSync('docker', ['stop', seleniumId]);
    });
  });
});


dind/onrun.sh
#!/bin/bash
#
# Copyright 2015 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Runs commands defined in the variable ONRUN, which is intepreted as
# an array of strings. We expect ONRUN to be declared in this fashion:
#
#     declare -x ONRUN="\"echo first\" \"echo second\""
#
# In a Dockerfile, a list of ENV declarations will
# construct such an ONRUN variable; for example:
#
#    ENV ONRUN "echo first"
#    ENV ONRUN $ONRUN "echo second"
#
# This produces the equivalent declaration.
#
# Each quote-delimited command will be executed in order.
#
# Last, any arguments to this script are executed in bash.
#

set -o errexit
set -o nounset
set -o pipefail

/usr/sbin/rsyslogd

# Convert the the ONRUN variable into an array.
eval COMMANDS=(${ONRUN:-})
NUM_COMMANDS=${#COMMANDS[@]}

# Put each command into this temporary file and execute it.
TEMP=$(mktemp).sh
trap "rm -rf ${TEMP}" EXIT
for (( I=0; I<${NUM_COMMANDS}; I++ )); do
  COMMAND="${COMMANDS[$I]}"
  logger -p local0.info "Executing ONRUN command ${COMMAND}"
  echo "${COMMAND}" > $TEMP
  chmod +x $TEMP
  if ! $TEMP; then
    logger -p local0.info \
      "ONRUN command $((I+1)) of ${NUM_COMMANDS} failed: ${COMMAND}"
    exit 1
  else
    logger -p local0.info \
      "ONRUN command $((I+1)) of ${NUM_COMMANDS} succeeded: ${COMMAND}"
  fi
done

# Even though we have an EXIT trap set, clean up eagerly.
rm -rf ${TEMP}

# Now run whatever arguments we were given.  Spawn a child process rather
# than 'exec' to make sure bash stays PID 1 to reap zombie children.
# TODO(aalexand): This may be insufficient to correctly handle docker stopping
# a container as bash won't propagate SIGTERM down.  Should we run a better
# init-like something?  See
# https://blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem/
/bin/bash -c "$*"


codiad/start.sh
# Copyright 2015 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

#!/bin/bash
# Changing the root's home dir to `/usr/share/nginx/www/_/workspace'.
export HOME=/usr/share/nginx/www/_/workspace
sed -i 's;:/root:;:/usr/share/nginx/www/_/workspace:;' /etc/passwd

# Setting up the IDE's work directories
mkdir -p /usr/share/nginx/www/_/workspace
mkdir -p /usr/share/nginx/www/_/data
ln -s /usr/share/nginx/www/_/workspace /
ln -s /usr/share/nginx/www/_/workspace /usr/share/nginx/www/
ln -s /usr/share/nginx/www/_/data /usr/share/nginx/www/

cp -r -n /root/. /usr/share/nginx/www/_/workspace/

chown root:www-data /usr/share/nginx/www -R

chmod a+rwX /usr/share/nginx/www/_

# Config git
git config --global user.email ${USER_EMAIL}
git config --global user.name ${USER_EMAIL}
git config --global alias.appraise '!/usr/local/bin/git-appraise'

wget http://storage.googleapis.com/shipshape-cli/shipshape -P /usr/local/bin/

chmod +x /usr/local/bin/*

shipshape --hot_start --tag prod-release-2015-10-21 $HOME &

mkdir -p /var/log/supervisor
/usr/bin/supervisord -c /etc/supervisord.conf


codiad/git-credential-gerrit.sh
# Copyright 2015 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

#!/bin/bash
#
# Copyright 2013 Google Inc. All Rights Reserved.
#
export CLOUDSDK_COMPONENT_MANAGER_DISABLE_UPDATE_CHECK=1

echo username=git
echo password=$("${CLOUD_SDK}/bin/gcloud" auth print-access-token)


dev-common/gcloud_docker_auth.sh
#!/bin/bash
#
# Copyright 2015 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Updates .dockercfg with credentials to ${GCLOUD_CONTAINER_SERVER}.
#

set -o errexit
set -o nounset
set -o pipefail

# Only authorize if we have credentials.
if gcloud --quiet auth print-access-token &> /dev/null; then
  gcloud --quiet docker --authorize-only --server="${GCLOUD_CONTAINER_SERVER}"
fi



ide-proxy/deploy-config/install.sh
# Copyright 2015 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

#!/bin/bash

# Variables defining the deployment target
PROJECT=$1
RELEASE=$2

VERSION=1

alias gcloud='gcloud --quiet'

# Local tag for the Docker image we deploy
LOCAL_DOCKER_IMAGE="google/ide-appengine"

function defaultModuleExists() {
  local project=$1
  local count=$(gcloud preview app modules list default --project $project 2>&1 \
    | grep -o "^default" | wc -l)
  if [[ $count -gt 0 ]]; then
    return 0
  else
    return 1
  fi
}

function deployDefaultModule() {
  local target_project=$1
  local tmp_dir=$(mktemp -d --tmpdir=$(pwd))
  echo
  echo "Default module doesn't exist, deploying default module now..."
  cat > ${tmp_dir}/app.yaml <<EOF
module: default
runtime: python27
api_version: 1
threadsafe: true
handlers:
- url: /
  mime_type: text/html
  static_files: hello.html
  upload: (.*html)
EOF

  cat > ${tmp_dir}/hello.html <<EOF
<html>
  <head>
    <title>Sample Hello-World Page.</title>
  </head>
  <body>
    Hello, World!
  </body>
</html>
EOF
  local status=1
  gcloud preview app deploy --force --quiet --project $target_project \
    $tmp_dir/app.yaml --version v1
  [[ $? ]] && status=0 || echo "Failed to deploy default module to $target_project"
  rm -rf $tmp_dir
  return $status
}

# If there is no release is specified then use the local image:
if [ -z $RELEASE ]; then
  echo "Deploying the local image: ${LOCAL_DOCKER_IMAGE}"
else
  IMAGE="gcr.io/developer-tools-bundle/ide-proxy:${RELEASE}"
  echo "Deploying image: ${IMAGE}"
  gcloud docker pull ${IMAGE}
  docker tag -f ${IMAGE} ${LOCAL_DOCKER_IMAGE}
fi

if ! defaultModuleExists $PROJECT; then
  if ! deployDefaultModule $PROJECT; then
    exit
  fi
fi

# Ensure that a private networks exists for IDE
if [ -z "$(gcloud --project=${PROJECT} compute networks list | grep codiad)" ]; then
  gcloud --project="${PROJECT}" compute networks create codiad --range "10.0.0.0/24"
fi

echo "from ${LOCAL_DOCKER_IMAGE}" > ./Dockerfile

gcloud --project="${PROJECT}" preview app deploy --set-default --force --version=${VERSION} ./app.yaml '--docker-build=local'

rm -f ./Dockerfile


ide-proxy/server.js
// Copyright 2015 Google Inc. All Rights Reserved.

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//     http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

var appengine = require('appengine');
var http = require('http');
var httpProxy = require('http-proxy');
var url = require('url');
var exec = require('child_process').exec;
var spawn = require('child_process').spawn;
var mkdirp = require('mkdirp');
var util = require('util');
var fs = require('fs');
var agent = new http.Agent({ maxSockets: Number.MAX_VALUE });

var indexFile = fs.readFileSync('index.html', {'encoding': 'utf8'});

// Scheduling a task for syncing the workspaces with the bucket in gsutil.
var SYNC_FREQUENCY_IN_MS = 120000 /* in milliseconds */ ;

// A map to store users and their IDE container info.
var db = {};

// Sends an HTML response.
function resHtml(res, html) {
  res.writeHead(200, {
    'Content-Type': 'text/html'
  });
  res.write(html);
  res.end();
}

// Sends an plain text response.
function resPlain(res, plain) {
  res.writeHead(200, {
    'Content-Type': 'text/plain'
  });
  res.write(plain);
  res.end();
}

//
// Create a proxy server with custom application logic
//
var proxyServer = httpProxy.createProxyServer({
  agent: agent
});
proxyServer.on('error', function(error, req, res) {
  var json;
  console.log('proxy error: ' + error);
  if (!res.headersSent) {
    res.writeHead(500, {
      'content-type': 'application/json'
    });
  }

  json = {
    error: 'proxy_error',
    reason: error.message,
    solution: 'Try again.'
  };
  res.end(JSON.stringify(json));
});

//
// Listen to the `upgrade` event and proxy the WebSocket requests as well.
//
proxyServer.on('upgrade', function(req, socket, head) {
  proxy.ws(req, socket, head);
});

// This variable indicates whether workspace sync must be stopped or not. When the VM is shutting
// down and after we receive the /_ah/stop signal, we set this variable to true so that no more sync
// is being initiated.
var stopSync = false;

// Syncs the workspace for the user identified with 'email'.
function syncWorkspace(email) {
  if (stopSync) {
    return;
  }
  // Logic to prevent multiple concurrent sync if one of them takes too long.
  if (db[email].isSyncInProgress) {
    return;
  }
  db[email].isSyncInProgress = true;
  var syncProcess;
  var user_workspace = [process.env.IDE_DATA_DIR, 'workspaces', email].join('/');
  if (db[email].inInitialSync) {
    mkdirp.sync(user_workspace);
    console.log('Running the initial workspace sync for: ' + email);
    // Runs an asyncronous process.
    // Copy from GCS bucket to IDE:
    syncProcess = spawn('gsutil', ['-m', 'rsync', '-c', '-r',
                                   process.env.IDE_BUCKET + '/workspaces/' + email,
                                   user_workspace]);
  } else {
    console.log('Running the workspace sync to GCS bucket for: ' + email);
    // Runs an asyncronous process.
    // Copy from IDE to GCS bucket:
    syncProcess = spawn('gsutil', ['-m', 'rsync', '-c', '-r', '-d', '-e',
                                   user_workspace,
                                   process.env.IDE_BUCKET + '/workspaces/' + email]);
  }
  syncProcess.stdout.pipe(process.stdout);
  syncProcess.stderr.pipe(process.stderr);
  syncProcess.on('exit', function(code) {
    db[email].isSyncInProgress = false;
    if (db[email].inInitialSync) {
      if (code === 0) {
        db[email].inInitialSync = false;
        console.log(util.format('Initial workspace sync is finished for %s with code %s.', email, code));
      } else {
        console.log(util.format('Error during the initial workspace sync for %s with code %s.', email, code));
      }
    } else {
      console.log(util.format('Workspace sync is finished for %s with code %s.', email, code));
    }
  });
  // If an error occures during the initial sycn we need to redo it.
  syncProcess.on('error', function(err) {
    db[email].isSyncInProgress = false;
    if (db[email].inInitialSync) {
      console.log(util.format('Error during the initial workspace sync for %s: %s.', email, JSON.stringify(err)));
    }
  });
}

function syncWorkspacesWithGCS() {
  if (stopSync) {
    return;
  }
  for (var email in db) {
    if (!(db[email].isSyncInProgress)) {
      syncWorkspace(email);
    }
  }
}

// This function tries to see if user associated with `email' has a workspace in GCS.
// If yes then the `yesCallback' will be called and otherwise `noCallback'.
function hasGCSWorkspace(email, yesCallback, noCallback) {
  var lsProcess = spawn('gsutil', ['ls', process.env.IDE_BUCKET + '/workspaces/' + email]);
  lsProcess.stdout.pipe(process.stdout);
  lsProcess.stderr.pipe(process.stderr);
  lsProcess.on('exit', function(code) {
    if (code === 0) {
      yesCallback();
    } else {
      noCallback();
    }
  });
}

// Run the workspace sync task.
setInterval(syncWorkspacesWithGCS, SYNC_FREQUENCY_IN_MS);

function recreateIDE(email) {
  if (db[email] && db[email].status === 'created') {
    var cid = db[email].container_id;
    exec(util.format('docker stop %s && docker rm %s', cid, cid),
         function(error, stdout, stderr) {
      console.log(util.format('Stopped container: %s. Output: %s. Error: %s.',
                              cid, stdout, stderr));
    });
  }
  tryCreateIDE(email);
}

function tryCreateIDE(email) {
  db[email].status = 'creating';
  console.log('Creating an instance of IDE for ' + email);
  // Create
  createIDE(email, function(response) {
    if (response.error === null) {
      db[email].status = 'created';
      console.log('Created IDE for %s. Container ID is: %s. Port is: %s',
                  email, db[email].container_id, db[email].container_port);
    } else {
      console.log('Creating an instance of IDE for %s faild. ' +
                  'Next attempt in 10 seconds.', email);
      setTimeout(function() {tryCreateIDE(email);}, 10000);
    }
  });
}

// Creates an IDE container for the user.
function createIDE(user, callback) {
  if (user) {
    var user_workspace = [process.env.IDE_DATA_DIR, 'workspaces', user].join('/');
    mkdirp.sync(user_workspace);
    var docker_command =
        'docker run --privileged -d -p 8080' + ' -v ' + user_workspace +
        ':/usr/share/nginx/www/_' + ' -e USER_EMAIL="' + user + '" ' + process.env.IDE_IMAGE;
    exec(docker_command, function(error, stdout, stderr) {
      if (error) {
        // Report errors for `docker run ...` command
        callback({
          'error': error,
          'stdout': stdout,
          'stderr': stderr
        });
      } else {
        var container_id = stdout.toString().trim();
        var docker_port_command = 'docker port ' + container_id + ' 8080';
        exec(docker_port_command, function(error, stdout, stderr) {
          if (error) {
            // Report errors for `docker port ...` command
            callback({
              'error': error,
              'stdout': stdout,
              'stderr': stderr
            });
          } else {
            // Extract the port portion of the `docker port ...` result
            var port_regex = /(.*):(.*)$/gm;
            var container_port = port_regex.exec(stdout.toString().trim())[2];
            db[user].container_id = container_id;
            db[user].container_port = container_port;
            db[user].status = 'created';
            callback({'error': null});
          }
        });
      }
    });
  } else {
    callback({
      'error': 'Required params are missing!'
    });
  }
}

var server = http.createServer(function(req, res) {
  appengine.middleware.base(req, res, function() {});
  var urlMap = url.parse(req.url, true);
  var email;
  if (urlMap.pathname === '/_ah/health') {
    resPlain(res, 'ok');
  } else if (urlMap.pathname === '/_ah/start') {
    stopSync = false;
  } else if (urlMap.pathname === '/_ah/stop') {
    stopSync = true;
  } else if (urlMap.pathname === '/-/status') {
    email = req.appengine.user.email;
    res.writeHead(200, {
      'Content-Type': 'application/json'
    });
    if (!db[email]) {
      db[email] = {};
    }
    res.end(JSON.stringify(db[email]));
  } else if (urlMap.pathname === '/-/recreate') {
    email = req.appengine.user.email;
    recreateIDE(email);
    res.writeHead(302, {'Location': '/'});
    res.end();
  } else {
    email = req.appengine.user.email;
    if (!db[email]) {
      db[email] = {};
      db[email].inInitialSync = true;
      hasGCSWorkspace(email, /* yes */ function() {}, /* no */ function() {
        // No need for initial sync
        db[email].inInitialSync = false;
      });
      syncWorkspace(email);
      tryCreateIDE(email);
      resHtml(res, indexFile);
    } else if (db[email].status === 'creating') {
      resHtml(res, indexFile);
    } else if (db[email].status === 'created') {
      if (db[email].inInitialSync) {
        resHtml(res, indexFile);
      } else {
        proxyServer.web(req, res, {
          target: 'http://localhost:' + db[email].container_port,
          headers: {
            'Connection': 'keep-alive'
          }
        });
      }
    }
  }
});

var port = 8080;
console.log('IDE Dashboard server listening on port: ' + port);
server.listen(port);
